### PR TITLE
CVS-46702 Demultiplexing & manual gathering documentation

### PR DESCRIPTION
**Pull request by dkalinowski**

### CODE REVIEW

File: docs/demultiplexing.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+There might be use cases where one custom node library is used to produce unpredictable number of batch. To achieve it, `demultiply_count` can be set to `0`. This indicates that pipeline supports any number of batch returned by custom node: `(1,X,N,C,H,W,...)` - where `X` is dynamic `demultiply_count`. OpenVINO&trade; Model Server is capable of interpreting such dynamic batch and is able to split outputs into dynamic number of pipeline branches. When using dynamic `demultiply_count` parameters, only one demultiplexer can exist in pipeline. Important to note - when batch 0 is returned, pipeline stops its execution and reports an error to gRPC/REST response.
+
+## Multiple demultiplexers
+Directed Acyclic Graph Scheduler is not limited to single demultiplexer node in one pipeline definition. Each demultiplexer node which has no manually specified gather node will be automatically gathered in `response` node - meaning each demultiplexer adds one new dimension equal to `demultiply_count` into all pipeline outputs shape. This must be taken into account when interpreting response data in client applications.
+
+![diagram](multiple_demultiplexers.svg)
+
+## Manual gathering
```
Comment (dtrawins) [inline docs/demultiplexing.md:82]: manually?

**Comment by dkalinowski** [comment]
TODO: after @atobiszei MR with gathering layout is merged, we should update this documentation according to changes applied

File: docs/demultiplexing.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
+# Demultiplexing in Directed Acyclic Graph Scheduler
+
+## Pipeline branching
+Directed Acyclic Graph Scheduler allows creating pipelines with optional parameter `demultiply_count: N` which adds ability to any node (`DL model` and `custom`) to slice outputs into `N` separate sub outputs and branch pipeline execution into `N` sub pipelines. Following nodes will be executed by event loop `N` times independently from each other and results will be gathered and  packed into one output just before sending a response. Additionally `gather_from_node: <node_name>` parameter can be used to manually specify to perform gathering at any point in Directed Acyclic Graph.
```
Comment (atobiszei) [inline docs/demultiplexing.md:4]: Maybe if we use verb "branch" for demultiplication then use "converge" for gathering?

File: docs/demultiplexing.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+## Basic demultiplexer example and metadata explaination
+This example contains 2 consecutive models:
+- Model A accepts one input named `input` with shape `(1,224,224)`, layout `NHW` and floating point precision (later referred as `(1,224,224) FP32`) and produces 2 outputs: `output_A (1,3,100,100) FP32` and `output_B (1,3,40,130) FP32`. Since this model is configured as demultiplexer, the second dimension is very important - it must be equal to `demultiply_count` parameter specified in config.json configuration file. This allows scheduler to split all outputs into smaller chunks (so called `demultiplexing` in diagram below) and spawn new pipeline branches.
```
Comment (atobiszei) [inline docs/demultiplexing.md:8]: Update for different shapes handling

File: docs/demultiplexing.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+- Model B accepts 2 inputs: `input_A (1,100,100) FP32` and `input_B (1,40,130) FP32`. Please note how metadata must be matched with previous model output after demultiplication (with second dimension removed after slicing). Since in most cases it is not possible to match metadata, OpenVINO&trade; Model Server allows creating custom nodes to perform output postprocessing (TODO: link to custom node doc). This model produces one input: `output (1,50) FP32`.
```
Comment (atobiszei) [inline docs/demultiplexing.md:9]: TODO

File: docs/demultiplexing.md (hunk 2 above)
Comment (atobiszei) [reply docs/demultiplexing.md:4]: `If no "gather_from_node" is specified then convering will be done automatically before sending the response.`
Maybe add

File: docs/demultiplexing.md [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+                {"pipeline_output_name": {"node_name": "Model_B_node",
+                                          "data_item": "output"}}
+            ]
+        }
+    ]
+}
+```
+
+## Dynamic demultiply_count parameter
+There might be use cases where one custom node library is used to produce unpredictable number of batch. To achieve it, `demultiply_count` can be set to `0`. This indicates that pipeline supports any number of batch returned by custom node: `(1,X,N,C,H,W,...)` - where `X` is dynamic `demultiply_count`. OpenVINO&trade; Model Server is capable of interpreting such dynamic batch and is able to split outputs into dynamic number of pipeline branches. When using dynamic `demultiply_count` parameters, only one demultiplexer can exist in pipeline. Important to note - when batch 0 is returned, pipeline stops its execution and reports an error to gRPC/REST response.
```
Comment (atobiszei) [inline docs/demultiplexing.md:75]: Modify for different shapes handling.

File: docs/demultiplexing.md (hunk 1 above)
Comment (atobiszei) [reply docs/demultiplexing.md:82]: Configuring gathering?

File: docs/demultiplexing.md [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            ]
+        }
+    ]
+}
+```
+
+## Pipeline configuration restrictions
+There are several restricted configurations in regards to demultiplexing and gathering:
+- You can gather only from nodes with `demultiply_count` specified (demultiplexer nodes)
+- Demultiplexer nodes and gathering nodes should be in LIFO order. Meaning you need to gather nodes starting from closest demultiplexer going upstream in Directed Acyclic Graph Scheduler
```
Comment (atobiszei) [inline docs/demultiplexing.md:165]:
```suggestion
- Demultiplexer nodes and gathering nodes should be in LIFO order. Meaning you need to gather nodes starting from closest demultiplexer going upstream in defined directed acyclic graph.
```
?

File: docs/demultiplexing.md (hunk 1 above)
Comment (dkalinowski) [reply docs/demultiplexing.md:82]: _Default_ word would fit better

Comment (dkalinowski) [reply docs/demultiplexing.md:82]: Add example use case

File: docs/demultiplexing.md [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+            "outputs": [
+                {"pipeline_output_name": {"node_name": "Model_C_node",
+                                          "data_item": "output"}}
+            ]
+        }
+    ]
+}
+```
+
+## Pipeline configuration restrictions
```
Comment (dkalinowski) [inline docs/demultiplexing.md:162]: Rules, not restrictions

**Comment by dkalinowski** [comment]
TODO: Add info about 0 results behavior, missing demulitplication at entry node

**Comment by dkalinowski** [comment]
TODO: Add info about restriction to only allow 1 demultiplexer when dynamic

File: docs/demultiplexing.md (hunk 2 above)
Comment (dkalinowski) [reply docs/demultiplexing.md:4]: "Results will be gathered" phrase is used here, "results will converge" sound odd. "Pipeline will converge" would  be better, but I'd like to stay with "results" here.

Comment (dkalinowski) [reply docs/demultiplexing.md:4]: `If no "gather_from_node" is specified then convering will be done automatically before sending the response.` - this is already implicated by phrase `Additionally `gather_from_node: <node_name>` parameter can be used to...`

File: docs/demultiplexing.md (hunk 6 above)
Comment (dkalinowski) [reply docs/demultiplexing.md:165]: All around the docs we capitalize first letters of DAG. Should we change it everywhere? @dtrawins

File: docs/demultiplexing.md (hunk 2 above)
Comment (dkalinowski) [reply docs/demultiplexing.md:4]: Additionaly, few lines below there is something similar:
```
This example has no `gather_from_node` parameter specified in config.json configuration file. This means, gathering (shown in diagram below) will be done automatically by `response` node at the end of event loop processing.
````

File: docs/demultiplexing.md (hunk 1 above)
Comment (dkalinowski) [reply docs/demultiplexing.md]: Done - got rid of `manual` word completely

File: docs/demultiplexing.md (hunk 3 above)
Comment (dkalinowski) [reply docs/demultiplexing.md]: Done

File: docs/demultiplexing.md (hunk 4 above)
Comment (dkalinowski) [reply docs/demultiplexing.md]: Added link

File: docs/demultiplexing.md (hunk 5 above)
Comment (dkalinowski) [reply docs/demultiplexing.md]: Done

File: docs/demultiplexing.md (hunk 7 above)
Comment (dkalinowski) [reply docs/demultiplexing.md]: Done

File: docs/demultiplexing.md (hunk 6 above)
Comment (atobiszei) [reply docs/demultiplexing.md]: Those capitals here looks to me that you are referring to the feature itself. In this case you are referring to one pipeline that you are defining - that's why in this context it looks odd especially with "Scheduler".

File: docs/demultiplexing.md [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+- You can gather only from nodes with `demultiply_count` specified (demultiplexer nodes)
+- When pipeline with dynamic `demultiply_count` encounters 0 results execution is stopped and gRPC/REST response returns specific error with such information
+- For pipelines with dynamic `demultiply_count` only 1 demultiplexer node is allowed.
+- Demultiplexer nodes and gathering nodes should be in LIFO order. Meaning you need to gather nodes starting from closest demultiplexer going upstream in Directed Acyclic Graph
+
+![diagram](demultiplexing_restriction_lifo.svg)
+
+- Node inputs must be coming from the same level of nested demultiplexing. To reduce demultiplexing levels you can gather results at any stage using `gather_from_node` parameter
+
+![diagram](demultiplexing_restriction_session_levels.svg)
```
Comment (dtrawins) [inline docs/demultiplexing.md:174]: We should include also a section with current limitations like input BS and models BS

File: docs/demultiplexing.md [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+                                          "data_item": "output"}}
+            ]
+        }
+    ]
+}
+```
+
+## Dynamic demultiply_count parameter
+There might be use cases where one custom node library is used to produce unpredictable number of batch. To achieve it, `demultiply_count` can be set to `0`. This indicates that pipeline supports any number of batch returned by custom node: `(X,N,C,H,W,...)` - where `X` is dynamic `demultiply_count`. OpenVINO&trade; Model Server is capable of interpreting such dynamic batch and is able to split outputs into dynamic number of pipeline branches. When using dynamic `demultiply_count` parameters, only one demultiplexer can exist in pipeline. Important to note - when batch 0 is returned, pipeline stops its execution and reports an error to gRPC/REST response.
+
```
Comment (dtrawins) [inline docs/demultiplexing.md:76]: stops execution and returns no content response.

**Review by dtrawins — APPROVED** [review]

File: docs/demultiplexing.md (hunk 9 above)
Comment (dkalinowski) [reply docs/demultiplexing.md:76]: Done

File: docs/demultiplexing.md (hunk 8 above)
Comment (dkalinowski) [reply docs/demultiplexing.md:175]: Done

File: docs/demultiplexing.md (hunk 6 above)
Comment (dkalinowski) [reply docs/demultiplexing.md]: Done

**Review by atobiszei — APPROVED** [review]

**Pull request merged**